lib/Core/Executor.cpp

```diff
@@ -905,14 +905,21 @@ Executor::fork(ExecutionState &current, ref<Expr> condition, bool isInternal) {
       }
     }
 
+    // We don't add a constraint explicitly yet
```

Comment:
@wuestholz , your understanding is correct.  Indeed, the question is when to effectively take scheduling decisions, and perhaps it would be worth revisiting this, making it consistent, and documenting it clearly.  There are many possibilities:
1) At every instruction
2) At every branch
3) At every (2-way) fork (what you call dynamic branch)
4) At every symbolic branch

We're essentially doing (1) now, so random would not get stuck.  I'm not sure whether it would make sense to switch states after every instruction in BFS though, but this is one interpretation of the strategy.

(2) and (3) are also worth considering.  (4) is a bit fragile (as it can be influenced by optimizations), so I'd drop it.  

I'll think a bit about it.  Do any of you have any (strong) preferences?